File: csharp/src/Fory/FlatMap.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Numerics;
using System.Runtime.CompilerServices;

namespace Apache.Fory;

internal static class TypeMapKey
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static ulong Get(Type type)
    {
        long value = type.TypeHandle.Value.ToInt64();
        return unchecked((ulong)value);
    }
}

/// <summary>
/// Open-addressed map keyed by <see cref="ulong"/>.
/// <para><see cref="ulong.MaxValue"/> is reserved as the empty-slot marker and must never be inserted.</para>
/// </summary>
internal sealed class UInt64Map<TValue>
{
    private const ulong EmptyKey = ulong.MaxValue;
    private const double DefaultLoadFactor = 0.5;
    private const ulong GoldenRatio = 0x9E3779B97F4A7C15;

    private struct Slot
    {
        public ulong Key;
        public TValue Value;
    }

    private Slot[] _entries;
    private int _tableCapacity;
    private int _mask;
    private int _shift;
    private int _count;
    private readonly double _loadFactor;
    private int _growThreshold;

    internal UInt64Map(int initialCapacity = 2, double loadFactor = DefaultLoadFactor)
    {
        _loadFactor = loadFactor;
        int capacity = NextPowerOfTwo(Math.Max(initialCapacity, 2));
        _entries = new Slot[capacity];
        _tableCapacity = capacity;
        _mask = capacity - 1;
        _shift = 64 - BitOperations.TrailingZeroCount((uint)capacity);
        _growThreshold = (int)(capacity * loadFactor);
        ClearEntries(_entries, capacity);
    }

    internal int Count => _count;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal bool TryGetValue(ulong key, out TValue value)
    {
        int index = Place(key);
        while (true)
        {
            ref Slot slot = ref _entries[index];
            ulong slotKey = slot.Key;
            if (slotKey == key)
            {
                value = slot.Value;
                return true;
            }

            if (slotKey == EmptyKey)
            {
                value = default!;
                return false;
            }

            index = (index + 1) & _mask;
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal void Set(ulong key, TValue value)
    {
        if (_count >= _growThreshold)
        {
            Grow();
        }

        int index = FindSlotForInsert(key);
        ref Slot slot = ref _entries[index];
        if (slot.Key == EmptyKey)
        {
            slot.Key = key;
            _count += 1;
        }

        slot.Value = value;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal bool Remove(ulong key)
    {
        int index = Place(key);
        while (true)
        {
            ref Slot slot = ref _entries[index];
            ulong slotKey = slot.Key;
            if (slotKey == key)
            {
                RemoveEntry(index);
                return true;
            }

            if (slotKey == EmptyKey)
            {
                return false;
            }

            index = (index + 1) & _mask;
        }
    }

    internal void Clear()
    {
        if (_count == 0)
        {
            return;
        }

        ClearEntries(_entries, _tableCapacity);
        _count = 0;
    }

    internal void ClearKeys()
    {
        if (_count == 0)
        {
            return;
        }

        ClearEntryKeys(_entries, _tableCapacity);
        _count = 0;
    }

    internal void AddValuesTo(List<TValue> values)
    {
        for (int i = 0; i < _tableCapacity; i++)
        {
            if (_entries[i].Key != EmptyKey)
            {
                values.Add(_entries[i].Value);
            }
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private int FindSlotForInsert(ulong key)
    {
        int index = Place(key);
        while (true)
        {
            ulong slotKey = _entries[index].Key;
            if (slotKey == EmptyKey || slotKey == key)
            {
                return index;
            }

            index = (index + 1) & _mask;
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private int Place(ulong key)
    {
        return (int)((key * GoldenRatio) >> _shift);
    }

    private void Grow()
    {
        Slot[] oldEntries = _entries;
        int oldCapacity = _tableCapacity;
        int newCapacity = oldCapacity << 1;
        Slot[] newEntries = new Slot[newCapacity];
        ClearEntries(newEntries, newCapacity);

        _entries = newEntries;
        _tableCapacity = newCapacity;
        _mask = newCapacity - 1;
        _shift = 64 - BitOperations.TrailingZeroCount((uint)newCapacity);
        _growThreshold = (int)(newCapacity * _loadFactor);
        _count = 0;

        for (int i = 0; i < oldCapacity; i++)
        {
            Slot oldSlot = oldEntries[i];
            if (oldSlot.Key != EmptyKey)
            {
                int newIndex = FindSlotForInsert(oldSlot.Key);
                _entries[newIndex] = oldSlot;
                _count += 1;
            }
        }
    }

    private void RemoveEntry(int index)
    {
        int hole = index;
        int cursor = (hole + 1) & _mask;

        while (_entries[cursor].Key != EmptyKey)
        {
            int idealSlot = Place(_entries[cursor].Key);
            if (ShouldMoveEntry(idealSlot, cursor, hole))
            {
                _entries[hole] = _entries[cursor];
                hole = cursor;
            }

            cursor = (cursor + 1) & _mask;
        }

        _entries[hole].Key = EmptyKey;
        _entries[hole].Value = default!;
        _count -= 1;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static bool ShouldMoveEntry(int idealSlot, int candidateSlot, int emptySlot)
    {
        if (emptySlot <= candidateSlot)
        {
            return idealSlot <= emptySlot || idealSlot > candidateSlot;
        }

        return idealSlot <= emptySlot && idealSlot > candidateSlot;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static void ClearEntryKeys(Slot[] entries, int count)
    {
        for (int i = 0; i < count; i++)
        {
            entries[i].Key = EmptyKey;
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static void ClearEntries(Slot[] entries, int count)
    {
        for (int i = 0; i < count; i++)
        {
            entries[i].Key = EmptyKey;
            entries[i].Value = default!;
        }
    }

    private static int NextPowerOfTwo(int value)
    {
        if (value <= 1)
        {
            return 1;
        }

        return 1 << (32 - BitOperations.LeadingZeroCount((uint)(value - 1)));
    }
}

File: swift/Sources/Fory/FlatMap.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

private let uint64MapEmptyKey = UInt64.max
private let uint64MapDefaultLoadFactor = 0.5
private let uint64MapGoldenRatio: UInt64 = 0x9E3779B97F4A7C15

/// Internal open-addressed UInt64-key map for runtime caches in Fory Swift
/// (for example type metadata caches in resolver/read/write contexts).
/// This is not part of the public API surface.
final class UInt64Map<Value> {

    private struct Slot {
        var key: UInt64
        var value: Value?
    }

    private var entries: [Slot]
    private var mask: Int
    private var shift: Int
    private var size = 0
    private let loadFactor: Double
    private var growThreshold: Int

    private var hasMaxKey = false
    private var maxKeyValue: Value?

    init(initialCapacity: Int = 2, loadFactor: Double = uint64MapDefaultLoadFactor) {
        self.loadFactor = loadFactor
        let capacity = Self.nextPowerOfTwo(max(initialCapacity, 2))
        let emptySlot = Slot(key: uint64MapEmptyKey, value: nil)
        entries = Array(repeating: emptySlot, count: capacity)
        mask = capacity - 1
        shift = UInt64.bitWidth - UInt64(capacity).trailingZeroBitCount
        growThreshold = Int(Double(capacity) * loadFactor)
    }

    var count: Int {
        size
    }

    var capacity: Int {
        entries.count
    }

    var isEmpty: Bool {
        size == 0
    }

    @inline(__always)
    func value(for key: UInt64) -> Value? {
        if key == uint64MapEmptyKey {
            return hasMaxKey ? maxKeyValue : nil
        }

        var index = place(key)
        while true {
            let entry = entries[index]
            if entry.key == key {
                return entry.value
            }
            if entry.key == uint64MapEmptyKey {
                return nil
            }
            index = (index + 1) & mask
        }
    }

    @inline(__always)
    func set(_ value: Value, for key: UInt64) {
        if key == uint64MapEmptyKey {
            if !hasMaxKey {
                hasMaxKey = true
                size += 1
            }
            maxKeyValue = value
            return
        }

        if size >= growThreshold {
            grow()
        }

        let index = findSlotForInsert(key)
        if entries[index].key == uint64MapEmptyKey {
            entries[index].key = key
            size += 1
        }
        entries[index].value = value
    }

    @inline(__always)
    func putIfAbsent(_ value: Value, for key: UInt64) -> (value: Value, inserted: Bool) {
        if key == uint64MapEmptyKey {
            if hasMaxKey, let maxKeyValue {
                return (maxKeyValue, false)
            }
            hasMaxKey = true
            maxKeyValue = value
            size += 1
            return (value, true)
        }

        if size >= growThreshold {
            grow()
        }

        let index = findSlotForInsert(key)
        if entries[index].key == uint64MapEmptyKey {
            entries[index].key = key
            entries[index].value = value
            size += 1
            return (value, true)
        }

        guard let existingValue = entries[index].value else {
            fatalError("corrupted UInt64Map state: occupied slot has nil value")
        }
        return (existingValue, false)
    }

    @discardableResult
    func removeValue(for key: UInt64) -> Value? {
        if key == uint64MapEmptyKey {
            guard hasMaxKey else {
                return nil
            }
            hasMaxKey = false
            size -= 1
            let removed = maxKeyValue
            maxKeyValue = nil
            return removed
        }

        var index = place(key)
        while true {
            let entry = entries[index]
            if entry.key == key {
                let removed = entry.value
                removeEntry(at: index)
                return removed
            }
            if entry.key == uint64MapEmptyKey {
                return nil
            }
            index = (index + 1) & mask
        }
    }

    func clear() {
        if !isEmpty {
            entries.withUnsafeMutableBufferPointer { buffer in
                guard let base = buffer.baseAddress else {
                    return
                }
                base.update(
                    repeating: Slot(key: uint64MapEmptyKey, value: nil),
                    count: buffer.count
                )
            }
            hasMaxKey = false
            maxKeyValue = nil
            size = 0
        }
    }

    @inline(__always)
    private func findSlotForInsert(_ key: UInt64) -> Int {
        var index = place(key)
        while entries[index].key != uint64MapEmptyKey && entries[index].key != key {
            index = (index + 1) & mask
        }
        return index
    }

    @inline(__always)
    private func place(_ key: UInt64) -> Int {
        Int((key &* uint64MapGoldenRatio) >> shift)
    }

    private func grow() {
        let oldEntries = entries
        let newCapacity = oldEntries.count * 2

        entries = Array(
            repeating: Slot(key: uint64MapEmptyKey, value: nil),
            count: newCapacity
        )
        mask = newCapacity - 1
        shift = UInt64.bitWidth - UInt64(newCapacity).trailingZeroBitCount
        growThreshold = Int(Double(newCapacity) * loadFactor)
        size = hasMaxKey ? 1 : 0

        for oldEntry in oldEntries where oldEntry.key != uint64MapEmptyKey {
            let newIndex = findSlotForInsert(oldEntry.key)
            entries[newIndex] = oldEntry
            size += 1
        }
    }

    private func removeEntry(at index: Int) {
        var hole = index
        var cursor = (hole + 1) & mask

        while entries[cursor].key != uint64MapEmptyKey {
            let idealSlot = place(entries[cursor].key)
            if shouldMoveEntry(idealSlot: idealSlot, candidateSlot: cursor, emptySlot: hole) {
                entries[hole] = entries[cursor]
                hole = cursor
            }
            cursor = (cursor + 1) & mask
        }

        entries[hole] = Slot(key: uint64MapEmptyKey, value: nil)
        size -= 1
    }

    @inline(__always)
    private func shouldMoveEntry(idealSlot: Int, candidateSlot: Int, emptySlot: Int) -> Bool {
        if emptySlot <= candidateSlot {
            return idealSlot <= emptySlot || idealSlot > candidateSlot
        }
        return idealSlot <= emptySlot && idealSlot > candidateSlot
    }

    private static func nextPowerOfTwo(_ value: Int) -> Int {
        if value <= 1 {
            return 1
        }
        var valueBits = UInt64(value - 1)
        valueBits |= valueBits >> 1
        valueBits |= valueBits >> 2
        valueBits |= valueBits >> 4
        valueBits |= valueBits >> 8
        valueBits |= valueBits >> 16
        if UInt64.bitWidth > 32 {
            valueBits |= valueBits >> 32
        }
        return Int(valueBits + 1)
    }
}
